routing/speed_camera_manager.cpp

```diff
@@ -0,0 +1,352 @@
+{
+std::string const SpeedCameraManager::kSpeedCamModeKey = "speed_cam_mode";
+
+SpeedCameraManager::SpeedCameraManager(turns::sound::NotificationManager & notificationManager)
+  : m_notificationManager(notificationManager)
+{
+  Reset();
+
+  uint32_t mode = 0;
+  if (settings::Get(kSpeedCamModeKey, mode) &&
+      mode < static_cast<int>(SpeedCameraManagerMode::MaxValue))
+  {
+    m_mode = static_cast<SpeedCameraManagerMode>(mode);
+  }
+  else
+  {
+    m_mode = SpeedCameraManagerMode::Auto;
+  }
+}
+
+//static
+SpeedCameraManager::Interval
+SpeedCameraManager::GetIntervalByDistToCam(double distanceToCameraMeters, double speedMpS)
+{
+  Interval interval;
+  if (distanceToCameraMeters < kInfluenceZoneMeters)
+  {
+    interval = Interval::ImpactZone;
+  }
+  else
+  {
+    double beepDist = kBeepSignalTime * speedMpS;
+    if (distanceToCameraMeters < kInfluenceZoneMeters + beepDist)
+      interval = Interval::BeepSignalZone;
+    else
+      interval = Interval::VoiceNotificationZone;
+  }
+
+  return interval;
+}
+
+void SpeedCameraManager::OnLocationPositionChanged(location::GpsInfo const & info)
+{
+  CHECK_THREAD_CHECKER(m_threadChecker, ());
+  CHECK(!m_route.expired(), ());
+
+  auto const passedDistanceMeters = m_route.lock()->GetCurrentDistanceFromBeginMeters();
+
+  // Step 1. Find new cameras and cache them.
+  FindCamerasOnRouteAndCache(passedDistanceMeters);
+
+  double distFromCurrentPosAndClosestCam = -1.0;
+  if (m_closestCamera.IsValid())
+  {
+    distFromCurrentPosAndClosestCam = m_closestCamera.m_distFromBeginMeters - passedDistanceMeters;
+    m_speedLimitExceeded = IsSpeedHigh(distFromCurrentPosAndClosestCam, info.m_speedMpS, m_closestCamera);
+  }
+  else
+  {
+    m_speedLimitExceeded = false;
+  }
+
+  // Step 2. Check cached cameras. Do it only after pass through closest camera.
+  if (!m_cachedSpeedCameras.empty() && distFromCurrentPosAndClosestCam < 0)
+  {
+    // Do not use reference here, because ProcessCameraWarning() can
+    // invalidate |closestSpeedCam|.
+    auto const closestSpeedCam = m_cachedSpeedCameras.front();
+
+    bool needUpdateClosestCamera = false;
+    auto const distToCamMeters = closestSpeedCam.m_distFromBeginMeters - passedDistanceMeters;
+
+    if (closestSpeedCam.m_distFromBeginMeters + kInfluenceZoneMeters > passedDistanceMeters)
+    {
+      needUpdateClosestCamera =
+        NeedUpdateClosestCamera(distToCamMeters, info.m_speedMpS, closestSpeedCam);
+
+      if (needUpdateClosestCamera)
+      {
+        m_closestCamera = closestSpeedCam;
+        ResetNotifications();
+        m_cachedSpeedCameras.pop();
+      }
+    }
+
+    if (NeedChangeHighlightedCamera(distToCamMeters, needUpdateClosestCamera))
+      PassCameraToUI(closestSpeedCam);
+  }
+
+  if (m_closestCamera.IsValid())
+    SetNotificationFlags(passedDistanceMeters, info.m_speedMpS, m_closestCamera);
+
+  // Step 3. Check UI camera (stop or not stop to highlight it).
+  if (m_currentHighlightedCamera.IsValid())
```

Comment:
nit: please, where it's the last instruction, use if-blocks in reverse fashion, like
``` 
if (!m_currentHighlightedCamera.IsValid())
  return;

auto const distToCameraMeters =
       m_currentHighlightedCamera.m_distFromBeginMeters - passedDistanceMeters;
if (!IsHighlightedCameraExpired(distToCameraMeters))
  return;

m_speedCamClearCallback();
m_currentHighlightedCamera.Invalidate();
```
it makes your code easy to read and there will be less meaningless braces, indents etc.

Reply:
yes, done

Reply:
tnx